doc: Add more application documentation (#211)
TheThingsIndustries/generic-node-se: Software/app/basic_bootloader/README.md

# basic_bootloader

## Description

The application provides a basic bootloader that can be combined with other applications to perform a specific functionality at boot time.

> > If you want to learn more about the purpose of a bootloader, we recommend reading this great [blog post](chttps://interrupt.memfault.com/blog/how-to-write-a-bootloader-from-scratch).

Reviewer: This does not seem to produce a clickable link for me, probably because of the stray 'c' before https (but double-check because I would expect it to take me to a non-existent website or something, but I can't interact at all which is weird).

Author: good catch

`basic_bootloader` performs one of two functionalities based on the GNSE button:

> 1. If the button is not pressed -> Boots the user flashed application
> 2. If the button is pressed long enough -> Boots ST internal bootloader

Reviewer 2: ```suggestion
1. If the button is not pressed -> boots the user flashed application
2. If the button is pressed long enough -> boots the ST internal bootloader
```

## Configuration

`basic_bootloader` key configurations:

### Memory Map

> The internal flash (ROM) memory map can be configured to accommodate a bootloader and an application by adjusting the parameters defined in [memory_map.ld](./../../target/memory_map.ld).

Reviewer 2: ```suggestion
The internal flash (ROM) memory map can be configured to accommodate a bootloader and an application by adjusting the parameters defined in [`memory_map.ld`](./../../target/memory_map.ld).
```

```
BOOTROM  (rx)  : ORIGIN = 0x08000000, LENGTH = 0x00000B000   /* Flash memory dedicated to bootloader */
APPROM   (rx)  : ORIGIN = 0x0800B000, LENGTH = 0x000035000   /* Flash memory dedicated to application */
```

### App activity

The application behavior can be adjusted by modifying [`conf/app_conf.h`](./conf/app_conf.h).

- `GNSE_TINY_TRACER_ENABLE` enables UART (115200/8-N-1) logging of application activity

```c
#define GNSE_TINY_TRACER_ENABLE 1
```

> > Please note: it is considered a best practice disable logging when building bootloaders to reduce the binary size.

Reviewer 2: ```suggestion
> Please note that it is considered as a best practice to disable logging when building bootloaders to reduce the binary size.
```

## Setup

> To build a `baic_bootloader` that can be combined with another application in the same memory, you need to use a linker script that supports the new memory map.

Reviewer: basic_bootloader

Author: good catch as well

### Building with CMake

> If you are building using CMake, then update [CMakeLists.txt](./../../CMakeLists.txt) `LINKER_SCRIPT` flag to `bootloader.ld` when compiling a `basic_bootloader` binary.

Reviewer 2: ```suggestion
If you are building using CMake, then update [`CMakeLists.txt`](./../../CMakeLists.txt) `LINKER_SCRIPT` flag to `bootloader.ld` when compiling a `basic_bootloader` binary.
```

> If you are building using CMake, then update [`CMakeLists.txt`](./../../CMakeLists.txt) `LINKER_SCRIPT` flag to `bootloader.ld` when compiling a `basic_bootloader` binary.

Reviewer: Maybe link for bootloader.ld could be added

Author: This is intentional, see this line
https://github.com/TheThingsIndustries/generic-node-se/blob/92cf8069113548399029a2c4555a5e38ed44e5d2/Software/CMakeLists.txt#L58

```
set(LINKER_SCRIPT ${MCU_DIR}/bootloader.ld)
```

Also, make sure to update `LINKER_SCRIPT` flag to `app.ld` when compiling an application binary that will be combined with the bootloader binary.

```
set(LINKER_SCRIPT ${MCU_DIR}/app.ld)
```

### Building with STM32CubeIDE

> If you are building using STM32CubeIDE, then there is no need to change the linker script when building a `basic_bootloader` binary as seen below.

Reviewer 2: ```suggestion
If you are building using STM32CubeIDE, then there is no need to change the linker script when building a `basic_bootloader` binary as can be seen below.
```

> If you are building using STM32CubeIDE, then there is no need to change the linker script when building a `basic_bootloader` binary as can be seen below.

Reviewer: Maybe add a small comment that you should right-click and select properties to get this window

Author: It's a good idea but I didn't want to go into Eclipse window management and navigation here.
I expect people who have trouble with this are the same people who shouldn't be changing linker scripts in the first place.

Reviewer: Good point

![basic_booloader STM32CubeIDE linker config](./../../../Docs/img_sw/stm32cubeide_linker_bootloader.png)

> But make sure to change the application linker script from `../../../target/stm32wl55xx_flash.ld` to `../../../target/app.ld`

Reviewer 2: ```suggestion
But, make sure to change the application linker script from `../../../target/stm32wl55xx_flash.ld` to `../../../target/app.ld`.
```

![basic_lorawan STM32CubeIDE linker config](./../../../Docs/img_sw/stm32cubeide_linker_app.png)

> 

Reviewer 2: ```suggestion
```

### Combining applications

> You can combine two binaries such as the `basic_bootloader` and `basic_lorawan` by using the [bootloader_app_merge python script](./../../bootloader_app_merge.py).

Reviewer 2: ```suggestion
You can combine two binaries such as the `basic_bootloader` and `basic_lorawan` by using the [`bootloader_app_merge` python script](./../../bootloader_app_merge.py).
```

For example:

```sh
python bootloader_app_merge.py \
-b ./build/debug/app/basic_bootloader/main.bin \
-a ./build/debug/app/basic_lorawan/main.bin \
-s 1 \
-d 0xB000 \
-c 0xB008 \
-o ./build/debug/app/basic_lorawan/boot_basic_lorawan.bin

> ```

Reviewer 2: add a blank line

## Observation

The device boots and turns on the red LED.

> If the button is not pressed, the red LED will stay on for a small period and the device will boot the user flashed application.

Reviewer 2: ```suggestion
If the button is not pressed, the red LED will stay on for a small period and the device will boot the user-flashed application.
```

> If the button is pressed for long enough it will turn on the blue LED and turn off the red LED indicating successful input and it continues to boot ST internal bootloader.

Reviewer 2: ```suggestion
If the button is pressed for long enough, it will turn on the blue LED and turn off the red LED, indicating a successful input and it will continue to boot the ST internal bootloader.
```
